### ISSUE TITLE
Proxy support

### ISSUE THREAD

**Issue by TyDzej** [AwesomeTTS/AwesomeTTS#90]
AwesomeTTS ignor enviromental settings of proxy it means that bypass proxy and all attemps to use online TTS service fail. How to enforce AwesomeTTS to use proxy?

**Comment by corpulentcoffee** [comment]
Most services use `urllib2` underneath, which I thought respected `HTTP_PROXY`/`HTTPS_PROXY` from the environment automatically.

I can take a look when I have time, though. What is your setup? (anything you can share would be helpful: operating system version, Anki version. Python version, type of proxy, which environment variable the proxy is advertised by)

**Comment by TyDzej** [comment]
I’ve captured packets through wireshark and all traffic is directed to destination directly not through proxy. Proxy in my environment is set through wpad.dat. Anki sync or AwesomeTTS plug-in works well but online TTS skip proxy directives. In company network is all traffic prohibited and we have to use proxy.
My configuration is Win 7 64-bit, but it did on 32-bit as well.

**Comment by corpulentcoffee** [comment]
I've never heard of wpad. Do you know if it makes its configuration available to regular programs via `HTTP_PROXY` or `HTTPS_PROXY` or something like that? Does other network functionality in Anki work? (like downloading add-ons?)

**Comment by corpulentcoffee** [comment]
(like, when you say that "AwesomeTTS plug-in works well", does anything it does online work, or do *all* online services fail?)

**Comment by TyDzej** [comment]
It does for sure, because Anki works in same way and there works everything (deck download, plug-in download, synchronization, plug-in upgrade, shared deck download). It just set my proxy to 10.x.x.x:8080. If I set proxy in system manually to something else Anki stops working.

Have I answered your question?

**Comment by corpulentcoffee** [comment]
Can you open Anki, press `Ctrl` + `Shift` + `:` and then enter `import os; print os.environ.get('HTTP_PROXY'), os.environ.get('HTTPS_PROXY')` and then press `Ctrl` + `Enter` and tell me what you get?

I'm going to look to see if `urllib2` needs to specially fed proxy information; I just need to make sure that I'm getting it from the right place to work on your system.

**Comment by TyDzej** [comment]
this is output:
>>> import os; print os.environ.get('HTTP_PROXY'), os.environ.get('HTTPS_PROXY')
None None

It doesn't look good. How Anki know that should use PROXY
If I capture packets during sync it goes through proxy .....
Any idea?

**Comment by corpulentcoffee** [comment]
Anki itself seems to be using a mix of `urllib2` calls (just like the ones AwesomeTTS uses) and `httplib2` calls (very unlike AwesomeTTS)... maybe the proxy support is scattered?

Make case makes a difference. Can you try:

    import os; print os.environ.get('http_proxy'), os.environ.get('https_proxy')

And then can you try the other library? Like this:

    import httplib2; print httplib2.proxy_info_from_environment()

Thanks, I appreciate it!

**Comment by TyDzej** [comment]
here we go
>>> import os; print os.environ.get('http_proxy'), os.environ.get('https_proxy')
None None

>>> import httplib2; print httplib2.proxy_info_from_environment()
None

It must be somewhere else, because when I set Proxy in Internet Explorer manualy, Anki stops wokr but output of your commands is same

**Comment by corpulentcoffee** [comment]
Thank you!

I'm going to play around with this in a virtual machine over the weekend. I
will try to reproduce it.

**Comment by corpulentcoffee** [comment]
Hey @TyDzej --

I was not able to duplicate the proxy problem in a virtual machine (everything I tried was picked up automatically by `urllib2` as implemented today). However, we try something else.

I tried hacking together this using `httplib2` and the `HTTP_PROXY` value from Anki's own sync mechanism: https://github.com/AwesomeTTS/AwesomeTTS/compare/proxy-90

If you want to try it, grab the zip from https://github.com/AwesomeTTS/AwesomeTTS/archive/proxy-90.zip. Close Anki, and then open the zip, navigate into `addon`, and copy `awesometts` and `AwesomeTTS.py` to your Anki add-on directory.

**Comment by TyDzej** [comment]
Hi Dave,

I'm not sure what should be the output. I replaced AwesomeTTS with the one from zip and run ANKI. I went to Configuration - Manage Presents and tried to play some word in several services - it doesn't help at all.

I suspect ANKI that it doesn't use PROXY explicitly but it call some other service to serve http requests and it doesn't go to internet directly. Is there some way how check this? Or set some debug?

**Comment by TyDzej** [comment]
Hi Dave,

I've found work around. In my case when I used http_proxy enviroment variable it didn't work and I found why. In my password was "@" and although I tried to substitute it with "%40" it didn't work. I changed special character to "?" what crash whole Anki :-)  but change it to "_" was what me works now.

To sum up now I can use AwesomeTTS in corporate network, but my plain text password is stored in environmental variable. If you figure out with my previous settings it would be appreciated (system has automatic proxy settings detection).

I’ll assist you if want.

Cheers Tom

**Comment by corpulentcoffee** [comment]
Hello!

First, after you unzip the package I linked, if you go to Tools > AwesomeTTS, what version string do you see in the top-right corner?

Second, the `HTTP_PROXY` value that is imported from `anki.sync` can be gotten from the registry. If you do `Ctrl` + `Shift` + `+` again and enter this:

    `import urllib; print urllib.getproxies_registry()`

.... I wonder what the output is?


I wonder if perhaps `urllib2` (which AwesomeTTS uses to do downloads) handles the `@` in your old password incorrectly but `httplib2` (which Anki uses for syncing) *does* handle it correctly.

Cheers!

**Comment by TyDzej** [comment]
Hi Dave,

I'm still on :-) It advertise AwesomeTTS v1.8.0-dev so I suppose it is correct.

>>> import urllib; print urllib.getproxies_registry()
{}

>>> import os; print os.environ.get('http_proxy'), os.environ.get('https_proxy')
None None

>>> import httplib2; print httplib2.proxy_info_from_environment()
None

I know IP address and canonical name of proxy server. Is it possible to do some dump of all variables and try to look it up in it?
Tom

**Comment by corpulentcoffee** [comment]
Yes, `1.8.0-dev` is what I was looking for.

You can get to all of your environment variables via `os.environ` (try something like `import os; print os.environ.items()`), but there is a ton of stuff in the Windows registry.

Does the regular version of AwesomeTTS from the add-ons site (version `1.7.0`) work for you now that your password doesn't have `@` or `?` in it?

Cheers!

**Comment by TyDzej** [comment]
Hi Dave,

I’m genuinely sorry. In os.environ was my proxy and I did packet capture again and I noticed that to IPs 178.63.105.134 and 148.251.54.24 goes my computer directly without proxy (these addresses are for anki sync, right?). What means that there is exception in our corporate policy and I don’t need proxy to use anki sync and other stuff. I was confused because Anki worked and AwesomeTTS didn’t (there isn’t exception). My former attempts to use proxy failed on “@” in password as you know.

I’m really sorry that I didn’t notice it before I bothered you.

Yes, v 1.7.0 works.
Cheers Tom
